Website 38 investigate adding the ability to navigate via calendar in hours view

SalazarJosh opened the pull request:
# Overview
Users can only navigate one week a time on the [Hours View](https://lib.umich.edu/locations-and-hours/hours-view), which makes it tedious to look further ahead. 

This PR adds a calendar component to the hours view ([deploy URL](https://deploy-preview-319--future-wwwlib-previews.netlify.app/locations-and-hours/hours-view#art-architecture-and-engineering-library)). The calendar component provides a singular component to give the user the ability to navigate by weeks while also having a monthly calendar display.

The calendar is a sticky header component that scrolls with the user. The previous implementation of having a previous / next week component for each section caused a significant amount of visual information. Hopefully this alleviates some of that.

ChatGPT helped write some of the algorithms for comparing JS `Date` objects, since those aren't always super intuitive. I gave it all an additional pass to clean up / combine logic and make it more human legible.

> This pull request resolves/closes/fixes [WEBSITE-38](https://mlit.atlassian.net/jira/software/projects/WEBSITE/boards/27?selectedIssue=WEBSITE-38).

## Anything else?
As part of this work, I was exploring anchor links since the hours view is the only page we support anchor links. For example: [https://deploy-preview-319--future-wwwlib-previews.netlify.app/locations-and-hours/hours-view#art-architecture-and-engineering-library](https://deploy-preview-319--future-wwwlib-previews.netlify.app/locations-and-hours/hours-view#art-architecture-and-engineering-library)

We handle anchor links in the `gatsby-browser` file. I extended the function to offset the anchor position by the height of the sticky header.

I also explored fixing the tab order for anchor links. In essence, when a user gets to a page via an anchor link, I want the anchored element to receive focus. Instead of trying to explore that issue within this PR, I broke it out into a separate ticket.

## Testing
List instructions on how to test the pull request. Some examples:

- Make sure the PR is consistent in these browsers:
  - [x] Chrome
  - [x] Firefox
  - [ ] Safari (the assignee was not able to test the pull request in this browser)
  - [x] Edge 
- Run accessibility tests:
  - [x] WAVE
  - [x] ARC Toolkit
  - [x] axe DevTools

erinesullivan commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,586 @@
+import PropTypes from 'prop-types';
+import { useStateValue } from '../use-state';
+
+const dateFormat = (string, abbreviated = false) => {
+  if (abbreviated) {
+    return string.toLocaleString('en-US', {
+      day: 'numeric',
+      month: 'short'
+    });
+  }
+  return string.toLocaleString('en-US', {
+    day: 'numeric',
+    month: 'long',
+    weekday: 'long',
+    year: 'numeric'
+  });
+};
```
Not a blocker, but an opportunity to reduce code:
```suggestion
const dateFormat = (date, abbreviated = false) => {
  const options = {
    day: 'numeric',
    month: abbreviated ? 'short' : 'long',
    ...(abbreviated ? {} : { weekday: 'long', year: 'numeric' })
  };
  
  return date.toLocaleString('en-US', options);
};
```

SalazarJosh replied:
+1 for less code! Always in favor of that. Thanks!

erinesullivan commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,586 @@
+  });
+};
+
+const IconWrapper = (props) => {
+  return (
+    <span
+      css={{
+        display: 'inline-block',
+        marginTop: '-2px'
+      }}
+      {...props}
+    />
+  );
+};
```
This component might not even be needed. They are only used in `PreviousNextWeekButton`, which is inside a `flex` element. You could add `align-items: center;` and `gap: SPACING['2XS']` to the parent, and remove the extra CSS used in the wrapped `Icon` components.

erinesullivan commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,586 @@
+
+  const toggleCalendarVisibility = () => {
+    setIsCalendarVisible(!isCalendarVisible);
+    setCalendarStatus(
+      isCalendarVisible ? 'Calendar closed' : 'Calendar open'
+    );
```
```suggestion
    setCalendarStatus(`Calendar ${isCalendarVisible ? 'closed' : 'open'}`);
```

erinesullivan commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,586 @@
+      <div
+        aria-live='polite'
+        aria-atomic='true'
+        style={{
+          clip: 'rect(0 0 0 0)',
+          clipPath: 'inset(50%)',
+          height: '1px',
+          overflow: 'hidden',
+          position: 'absolute',
+          whiteSpace: 'nowrap',
+          width: '1px'
+        }}
```
Is this to visually hide the element? If so, there is the `visually-hidden` class.
```css
.visually-hidden {
    position: absolute !important;
    height: 1px;
    width: 1px;
    overflow: hidden;
    clip: rect(1px 1px 1px 1px);
    clip: rect(1px, 1px, 1px, 1px);
}
```

erinesullivan commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,586 @@
+            >
+              {['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'].map((day) => {
+                return (
+                  <div
+                    css={{
+                    }}
+                    key={day}
+                  >
```
Any reason for the blank `css` attribute?
```suggestion
                  <div key={day}>
```

erinesullivan commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,586 @@
+                    key={weekIndex}
+                    aria-label={`${currentlySelectedWeek ? 'This is the currently selected week.' : 'View'} week ${weekIndex + 1} in ${month} from ${getFormattedDate(week[0])} to ${getFormattedDate(week[week.length - 1])}`}
+                    css={{
+                      border: currentlySelectedWeek
+                        ? `2px solid var(--color-maize-400)`
+                        : `2px solid rgba(0,0,0,0)`,
```
```suggestion
                      border: `2px solid ${currentlySelectedWeek ? 'var(--color-maize-400)' : 'transparent'}`,
```

erinesullivan commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,586 @@
+import { Heading, Icon, Link, Margins, MEDIA_QUERIES, SPACING, TYPOGRAPHY } from '../../reusable';
```
```suggestion
import { Heading, Icon, Margins, MEDIA_QUERIES, SPACING, TYPOGRAPHY } from '../../reusable';
```
ESLint error: `1:25  error  'Link' is defined but never used  no-unused-vars`

erinesullivan commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,586 @@
+              })}
+            </div>
+            <p css={{ color: `var(--color-neutral-300);` }}>
+              <span css={{ background: 'white',
+                border: `solid 1px var(--color-neutral-200)`,
+                borderRadius: '4px',
+                boxShadow: `0 1px 1px rgba(0, 0, 0, .2), 0 2px 0 0 rgba(255, 255, 255, .7) inset;`,
+                display: 'inline-block',
+                fontFamily: 'monospace',
+                fontSize: '0.85rem',
+                marginBottom: SPACING.S,
+                padding: `0 ${SPACING['2XS']}` }}
+              >
```
I know that the `kbd` element is used in `src/components/site-search.js`. Might be a good opportunity to create a reusable component?

erinesullivan commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,586 @@
+  };
+
+  const handleInteraction = (event, action) => {
+    if (event.type === 'click' || (event.type === 'keydown' && (event.key === 'Enter' || event.key === ' '))) {
```
This same logic is also used in line 42: `event.type === 'click' || (event.type === 'keydown' && (event.key === 'Enter' || event.key === ' '))`

Might be a good opportunity to create a function?

erinesullivan commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,586 @@
+          Previous week
+        </PreviousNextWeekButton>
+        <button
+          css={isCalendarVisible
+            ? {
+                boxShadow: 'inset 0 -2px var(--color-teal-400);',
+                color: 'inherit'
+              }
+            : {
+                '&:hover': {
+                  boxShadow: 'inset 0 -2px var(--color-teal-400);'
+                },
+                boxShadow: 'none',
+                color: 'inherit'
+              }}
```
Another opportunity to reduce code:
```suggestion
          css={{
  boxShadow: 'none',
  color: 'inherit',
  ...(isCalendarVisible
    ? {
        boxShadow: 'inset 0 -2px var(--color-teal-400)'
      }
    : {
        '&:hover': {
          boxShadow: 'inset 0 -2px var(--color-teal-400)'
        }
      })
}}
```

SalazarJosh reviewed (commented):
Thanks for taking the time to review! I've made all the suggested changes. Since your last review, Emma and Heidi got some user feedback and we made two more UI edits. I'll mark those so you don't have to review all the code again. Thanks again!

SalazarJosh commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,553 @@
+    >
+      <div
+        css={{
+          alignItems: 'center',
```
center aligned all items in the HoursPanelNextPrev component

SalazarJosh commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -0,0 +1,553 @@
+                        <div
+                          key={dayIndex}
+                          css={{
+                            backgroundColor: getColorBasedOnDate(day),
```
I was dynamically setting the color here if the day didn't fall in the current month. UX team found the color wasn't accessible on gold background (which is the case for last / first week of month if there are overlapping days. The suggestion was to make all the numbers the same color so that rule is now gone.

erinesullivan reviewed (changes requested):
This looks great! I have done accessibility testing, and there are two blockers:

- ✅ Everything works with keyboard and mouse.
- ❔ *NOT A BLOCKER!* I did VoiceOver testing and this is more of a question for Ben and Bridget on how descriptive should this be:
  - For the previous/next week/month buttons, they say the text. Would it be worth to voice what the next week or month the user will be moved into?
  - The calendar does not indicate whether it's opened or closed. Have you thought about trying to use the `details` element? I believe that's natively baked in.
- ❌ The `esc to close` text color does not have a high enough contrast.
- ❌ ARC Toolkit gave out this error: `ARIA attribute is not allowed` and this warning: `Missing href`. I believe these can be resolved in two ways:
  1. Continue using the anchor element, and add an empty `href` attribute, and move the `aria-label` value inside the anchor and wrap it in a `span.visually-hidden` element. You make also need to add `aria-hidden` to the days of the week so they're not voiced.
  2. Change the anchor to a button.

You're almost there!

SalazarJosh commented:
> This looks great! I have done accessibility testing, and there are two blockers:
> 
> * ✅ Everything works with keyboard and mouse.
> * ❔ _NOT A BLOCKER!_ I did VoiceOver testing and this is more of a question for Ben and Bridget on how descriptive should this be:
>   
>   * For the previous/next week/month buttons, they say the text. Would it be worth to voice what the next week or month the user will be moved into?
>   * Have you thought about trying to use the `details` element? I believe the open/close voicing is natively baked in.
> * ❌ The `esc to close` text color does not have a high enough contrast.
> * ❌ ARC Toolkit gave out this error: `ARIA attribute is not allowed` and this warning: `Missing href`. I believe these can be resolved in two ways:
>   
>   1. Continue using the anchor element, and add an empty `href` attribute, and move the `aria-label` value inside the anchor and wrap it in a `span.visually-hidden` element. You make also need to add `aria-hidden` to the days of the week so they're not voiced.
>   2. Change the anchor to a button.
> 
> You're almost there!

Thanks again! I made quite a few changes but I'll summarize here:
- I added an aria-live to the date range title so the previous / next week buttons announce changes on input. It's after the action versus before, but it's definitely still an improvement. I'll share with Ben / Bridget to get their input as well.
- Changed the 'to close' text in 'esc to close' to the neutral-400 color.
- I think the `details` element is a step in the right direction. After implementing it I was able to remove all the logic around tracking and setting calendar visibility. However, there are still quite a few bugs with the element, specifically in FireFox. Here's a [good writeup on it](https://www.scottohara.me/blog/2022/09/12/details-summary.html) from a couple of years ago.
- Fixed the Arc errors by changing the `<a>` to a `button` element. That seemed like the path of least resistance and was still clear when using NVDA screen reader.

SalazarJosh commented on src/components/panels/hours-panel-date-viewer.js:
```diff
@@ -316,7 +316,7 @@ const CalendarView = ({ isVisible, weekOffset }) => {
                 });
                 return (
-                  <a
+                  <button
```
@erinesullivan This worked quite well in NVDA. I'm curious if it's still functional and makes sense with VoiceOver.

erinesullivan reviewed (approved):
👍🏻 The accessibility issues have been resolved! This looks great.
